fix(lib): infinite scroll in items index page

antontranelis opened the pull request:
Lazy loading items on Index Page

antontranelis committed:
infinite scroll in items index page

mahula committed:
Merge branch 'main' into lazy-loading-items-index

mahula committed:
Merge branch 'main' into lazy-loading-items-index

mahula committed:
Merge branch 'main' into lazy-loading-items-index

antontranelis committed:
Merge branch 'main' into lazy-loading-items-index

antontranelis committed:
Merge branch 'main' into lazy-loading-items-index

antontranelis committed:
Merge branch 'main' into lazy-loading-items-index

antontranelis committed:
Merge branch 'main' into lazy-loading-items-index

mahula committed:
Merge branch 'main' into lazy-loading-items-index

mahula committed:
Merge branch 'main' into lazy-loading-items-index

mahula commented:
augment review

mahula committed:
Merge branch 'main' into lazy-loading-items-index

mahula committed:
fix(lib): use stable item id as React key instead of array index in OverlayItemsIndexPage

mahula committed:
fix(lib): correct misleading 'once' comment on filtering logic in OverlayItemsIndexPage

mahula committed:
fix(lib): replace ineffective isLoadingMore state with ref-based guard for infinite scroll

mahula committed:
test(lib): add unit tests for lazy loading pagination logic

- Extract filterSortAndPaginate pure function for testability
- Add 14 Vitest unit tests covering layer filtering, tag filtering, date sorting, and pagination
- Refactor OverlayItemsIndexPage to use extracted helper

mahula committed:
test(lib): add component tests for OverlayItemsIndexPage infinite scroll

- Mock child components and hooks to isolate scroll orchestration
- Mock IntersectionObserver with controllable callback
- Test initial batch rendering (30 items + sentinel)
- Test scroll triggers load-more via IntersectionObserver
- Test sentinel removal when all items loaded
- Test ref-based guard prevents duplicate batch loads
- Add data-testid to sentinel element for stable querying
- Remove redundant comments from earlier changes

mahula commented:
augment review

mahula commented on lib/src/Components/Templates/OverlayItemsIndexPage.tsx:
```diff
@@ -66,6 +70,45 @@ export const OverlayItemsIndexPage = ({
+    return () => {
+      observer.disconnect()
+    }
+  }, [hasMore, visibleItems.length])
```
By design and correctly guarded. Yes, the effect re-runs when `visibleItems.length` changes, which disconnects/reconnects the observer. If the sentinel is still in the viewport, the observer *will* fire again — but `isLoadingMoreRef.current` is `true` at that point (set synchronously in the callback), so the guard blocks it. The ref is only reset to `false` in the separate `useEffect` that runs after `visibleItems.length` changes, which means the next batch can then load. This is the intended "load until the sentinel scrolls out of view" behavior. No action needed.

mahula commented on lib/src/Components/Templates/OverlayItemsIndexPage.tsx:
```diff
@@ -66,6 +70,45 @@ export const OverlayItemsIndexPage = ({
+
+    const observer = new IntersectionObserver(
+      (entries) => {
+        const entry = entries[0]
```
The IntersectionObserver spec guarantees the callback is invoked with at least one entry when observing a single element. An empty `entries` array is not a real-world scenario. Adding a guard would be dead code — skipping this.

mahula commented on lib/src/Components/Templates/OverlayItemsIndexPage.spec.tsx:
```diff
@@ -0,0 +1,168 @@
+}
+
+function triggerIntersection() {
+  observerCallback(
```
Good point. Added `expect(mockObserve).toHaveBeenCalled()` after each `renderPage()` call in the 3 tests that use `triggerIntersection()`. This asserts the IntersectionObserver was actually set up before we invoke its callback, guarding against future refactors that might change effect timing.

mahula committed:
test(lib): assert observer setup before triggering intersection

- Add expect(mockObserve).toHaveBeenCalled() in 3 tests that use triggerIntersection()
- Guards against future refactors that might change effect timing

ulfgebhardt committed:
Merge branch 'main' into lazy-loading-items-index

ulfgebhardt reviewed (approved):
I looked at the code and that is good
